File: Program's_Contributed_By_Contributors/C++/sudoku_solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board)
    {
        solve(board);
    }
    bool solve(vector<vector<char>>& board) {
        for(int i=0;i<board.size();i++)
        {
            for(int j=0;j<board[0].size();j++)
            {
                if(board[i][j]=='.')
                {
                    for(char c='1';c<='9';c++)
                    {
                        if(isValid(board,i,j,c))
                        {
                            board[i][j]=c;
                        if(solve(board)==true)
                        {
                          return true;  
                        }
                        else
                        {
                            board[i][j]='.';
                        }
                    }
                }
                    return false;
                }
            }
        }
        return true;
    }
    bool isValid(vector<vector<char>>& board,int row,int column,char c)
    {
        for(int i=0;i<9;i++)
        {
            if(board[i][column]==c)
            {
                return false;
            }
            if(board[row][i]==c)
            {
                return false;
            }
            if(board[3*(row/3)+(i/3)][3*(column/3)+(i%3)]==c)
            {
                return false;
            }
        }
            return true;
    }
};

File: Program's_Contributed_By_Contributors/Python_Programs/sudoku_solver.py
def solve_sudoku(grid):
    # Check if the grid is already solved
    if is_solved(grid):
        return grid

    # Find the next empty cell
    row, col = find_empty_cell(grid)

    # Try each possible value for the empty cell
    for value in range(1, 10):
        if is_valid(grid, row, col, value):
            grid[row][col] = value
            solve_sudoku(grid)
            if is_solved(grid):
                return grid
            grid[row][col] = 0

    # If no solution was found, return None
    return None

def is_solved(grid):
    # Check if all rows, columns, and boxes are filled
    for row in range(9):
        for col in range(9):
            if grid[row][col] == 0:
                return False
    return True

def find_empty_cell(grid):
    # Find the first empty cell in the grid
    for row in range(9):
        for col in range(9):
            if grid[row][col] == 0:
                return row, col
    return None

def is_valid(grid, row, col, value):
    # Check if the value is valid for the given cell
    for i in range(9):
        if grid[row][i] == value or grid[i][col] == value:
            return False
    return True